internal/mesh/internal/mappers/workloadselectionmapper/workload_selection_mapper.go

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package workloadselectionmapper
+
+import (
+	"context"
+
+	"google.golang.org/protobuf/proto"
+
+	"github.com/hashicorp/consul/internal/controller"
+	"github.com/hashicorp/consul/internal/mesh/internal/mappers/common"
+	"github.com/hashicorp/consul/internal/resource/mappers/selectiontracker"
+	"github.com/hashicorp/consul/lib/stringslice"
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v2beta1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+// WorkloadSelecting denotes a resource type that uses workload selectors.
+type WorkloadSelecting interface {
+	proto.Message
+	GetWorkloads() *pbcatalog.WorkloadSelector
+}
+
+type Mapper[T WorkloadSelecting] struct {
+	workloadSelectionTracker *selectiontracker.WorkloadSelectionTracker
+	computedType             *pbresource.Type
+}
+
+func New[T WorkloadSelecting](computedType *pbresource.Type) *Mapper[T] {
+	if computedType == nil {
+		panic("computed type is required")
+	}
+	return &Mapper[T]{
+		workloadSelectionTracker: selectiontracker.New(),
+		computedType:             computedType,
+	}
+}
+
+// MapToComputedType is responsible for mapping types with workload selectors to the corresponding computed type
+// resources which are name-aligned with the workload. This function will also track workload selectors with the ids
+// from the workload-selectable types in the mapper.
+func (m *Mapper[T]) MapToComputedType(ctx context.Context, rt controller.Runtime, res *pbresource.Resource) ([]controller.Request, error) {
+	var zero T
+	data := zero.ProtoReflect().New().Interface().(T)
+	// check that there is data to unmarshall
+	if res.Data != nil {
+		if err := res.Data.UnmarshalTo(data); err != nil {
+			return nil, err
+		}
+	}
```

Comment:
Can't you use `resource.Decode[T]` here?

Reply:
Yep that worked! I didn't think it would but I guess it makes sense because this type has the superset of methods of `proto.Message` used in `resource.Decode`.